File: day8/day8.py
from collections import defaultdict
import functools
from math import lcm

def parse(data):
    path = data[0].strip()
    ends_with_a = []

    mp = defaultdict(tuple)
    for line in data[2:]:
        k, v = line.split(" = ")
        l, r = v[1:-2].split(",")
        mp[k] = (l.strip(), r.strip())
        if k.endswith("A"): ends_with_a += [k]

    return path, mp, ends_with_a

def path_to_zz(mp, cur, path):
    cnt = 0
    idx = 0
    while not cur.endswith("Z"):
        dir = path[idx]
        if dir == "L":
            cur = mp[cur][0]
        else:
            cur = mp[cur][1]

        cnt += 1
        idx += 1
        if idx == len(path):
            idx = 0
    return cnt

def solution(data):
    path, mp, _ = parse(data)
    return path_to_zz(mp, "AAA", path)

def gcd(a, b):
    if b == 0: return a
    return gcd(b, a % b)

def lcm(*args):
    def lcmi(a, b):
        return (a * b) // gcd(a, b)
    return functools.reduce(lcmi, args)

def solution2(data):
    path, mp, ends_with_a = parse(data)
    costs = [path_to_zz(mp, item, path) for item in ends_with_a]
    return lcm(*costs)

if __name__ == "__main__":
    data = [line for line in open("input1.txt", "r")]
    print(solution(data))
    print(solution2(data))

File: day8/day8.cpp

#include "../stdc++.h"
#include "../stringutils.h"

using namespace std;

unordered_map<string, vector<string>> mp;
vector<string> start_nodes;
string path;

int path_to_z(string cur, const string& path, unordered_map<string, vector<string>>& mp) {
    int cnt = 0;
    int idx = 0;

    while (cur.back() != 'Z') {
        char dir = path[idx];
        cur = mp[cur][(dir == 'R')];
        cnt += 1;
        idx = (idx + 1) % path.size();
    }
    return cnt;
}

void solution(const vector<string>& input) {
    cout << path_to_z("AAA", path, mp) << endl;
}

uint64_t gcd(uint64_t a, uint64_t b) {
    if (a == 0) return b;
    return gcd(b, a % b);
}

uint64_t lcm(uint64_t a, uint64_t b) {
    return a * b / gcd(a, b);
}

void solution2(const vector<string>& input) {
    auto result = accumulate(start_nodes.begin(), start_nodes.end(), 1LL, [](uint64_t acc, string node) {
        return lcm(acc, path_to_z(node, path, mp));
    });

    cout << result << endl;
}

int main(int argc, char** argv) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    ifstream file(argv[1]);

    if (!file.is_open()) {
        cout << "File not found" << endl;
        return 0;
    }

    vector<string> input;
    string line;
    while (getline(file, line)) {
        input.push_back(line);
    }
    file.close();

    path = strip(input[0], ' ');

    for (int i=2; i<input.size(); i++) {
        vector<string> tokens = split(input[i], '=');
        string key = strip(tokens[0], ' ');
        vector<string> values = split(tokens[1].substr(1, tokens[1].size() - 2), ',');

        for (auto v: values) {
            mp[key].push_back(strip(v, ' '));
        }

        if (key.back() == 'A') {
            start_nodes.push_back(key);
        }
    }

    solution(input);
    solution2(input);

    return 0;
}